### PR TITLE
Arcore with poly

### PR DESCRIPTION
**Pull request by Dorvaryn**
This is a simple addition of poly to load a model on the fly instead of loading a pre-packaged asset from resources. 

screenshot_20180501-111116

This PR is for review. It will be merged into a new folder once reviewed to avoid making the original sample too complex.

### CODE REVIEW

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                listener.onError(e)
+            }
+        })
+    }
+
+    private fun findCompatibleAssetFormat(assets: List<APIAsset>): Pair<APIAsset, APIFormat>? {
+        return assets.fold(null, ::bar)
+    }
+
+    private fun bar(acc: Pair<APIAsset, APIFormat>?, asset: APIAsset): Pair<APIAsset, APIFormat>? {
```
Comment (lgvalle) [inline ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt]: `bar` 👀

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/APIAssets.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.novoda.spikes.arcore.poly
+
+data class APIAssets(val assets: List<APIAsset>)
```
Comment (takecare) [inline ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/APIAssets.kt]: can these be `Api*` (same for others), so it's in sync with `PolyApi`

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/FormatDownloader.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    private fun downloadFile(apiFile: APIFile, success: (File) -> Unit, failure: (Exception) -> Unit, over: () -> Unit) {
+        val request = Request.Builder()
+                .url(apiFile.url)
+                .build()
+
+        client.newCall(request).enqueue(object : Callback {
+            override fun onResponse(call: Call, response: Response) {
+                if (state != State.STATE_DOWNLOADING) return
```
Comment (takecare) [inline ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/FormatDownloader.kt]: `{ }`?

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                .url(urlForKeywords(keywords))
+                .build()
+
+        val response = client.newCall(request).enqueue(object : Callback {
+            override fun onResponse(call: Call, response: Response) {
+                val adapter = moshi.adapter(APIAssets::class.java)
+                var assets: List<APIAsset>? = null
+                try {
+                    val body = response.body()?.string()
+                    Log.e("WTF", "Got " + body)
```
Comment (takecare) [inline ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt]: `Log.wtf()`? 😄

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+import com.squareup.moshi.Moshi
+import okhttp3.*
+import java.io.IOException
+
+class PolyApi {
+
+    companion object {
+        private const val HOST = "poly.googleapis.com"
+        private const val VERSION = "v1"
+        private const val API_KEY = "API_KEY_HERE"
```
Comment (takecare) [inline ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt]: seems like a nice use case for https://github.com/novoda/gradle-build-properties-plugin 😄

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt (hunk 1 above)
Comment (Dorvaryn) [reply ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt]: Woops ^^

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/APIAssets.kt (hunk 2 above)
Comment (Dorvaryn) [reply ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/APIAssets.kt]: Can do

File: ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt (hunk 5 above)
Comment (Dorvaryn) [reply ARCore/app/src/main/java/com/novoda/spikes/arcore/poly/PolyApi.kt]: Might be overkill here, it's just a sample

**Comment by Dorvaryn** [comment]
@lgvalle Yep the sequence is correct, it could be worth having a default but since this logic changes a bit with the next PR maybe not worth doing yet ?

**Comment by lgvalle** [comment]
Agree, not worth doing it if the next PR changes it 👍

**Comment by Dorvaryn** [comment]
The merge will be in https://github.com/novoda/spikes/pull/384

**Pull request closed without merging**